thrift: adding xDS RDS support for thrift proxy.

tkovacs-2 opened the pull request:
Commit Message:
Adding xDS RDS support for thrift proxy.

Additional Description:
- The current RDS implementation is tied to the HTTP protocol. To make it reusable for other protocols
  base classes containing the generic part are extracted from the existing code.
- The location of the base classes are envoy/rds and source/common/rds and the Envoy::Rds namespace.
- Existing code is reworked to use the base classes and HTTP specific features added back by decorator pattern
  (StaticRouteConfigProviderImpl, RdsRouteConfigProviderImpl, RouteConfigUpdateReceiverImpl)
  and template method pattern (RdsRouteConfigSubscription).
- The RDS implementation for Thrift is simply the direct usage of base classes.
- The only thing which has to be implemented separately for every protocol is the
  RouteConfigProviderManagerImpl::createRdsRouteConfigProvider and
  RouteConfigProviderManagerImpl::createStaticRouteConfigProvider functions.
  To make it easier helper functions are added to Rds::RouteConfigProviderManagerImpl.
- New grpc/restapi endpoint is added for Thrift route discovery service. 

Risk Level:
Medium

Testing:
- Unittest
- Manual test with static routing from bootstrap config, static routing and dynamic routing from GRPC based xDS server.

Docs Changes:
Comment added to the new API field.

Release Notes:

Platform Specific Features:

rgs1 commented:
@tkovacs-2 this is great, thank you!  Any reasons to avoid base classes in source/common/router/ ? My intuition is that we'd want to share, unless you think there'll be irreconcilable differences between HTTP and Thrift towards handling RDS. 

cc: @htuch to get more guidance towards sharing common RDS code vs having Thrift and HTTP diverge.

rgs1 commented on test/extensions/filters/network/thrift_proxy/rds_impl_test.cc:
```diff
@@ -0,0 +1,582 @@
+  - match:
+      method_name: bar
+    route:
+      cluster: baz
```
nit: maybe add a mirroring policy here (#17544)

rgs1 reviewed (commented):
Generally LGTM, besides my previous comment about possibly adding base classes to share code with the existing router RDS implementation.

htuch commented:
I'm not sure I fully understand what is proposed. How can Thrift make use of RDS, which is about HTTP routes?

We definitely should not fork the RDS code, there is a lot of subtelty and complexity in the warming, shared providers and lifecycle (for on-demand). I would recommend making the base code more generic if it doesn't add too much complexity to Envoy core (i.e. it's just about making some resource type opaque).

tkovacs-2 commented:
@htuch @rgs1 
Thank you for your comments.

If the static routing rules change in a listener a new listener is created and after a drain period the old one is deleted. This has two negative consequences:
- During the drain period the clients connected to old listener are using the outdated routing rules.
- When the drain period is over the connections will be closed by envoy which can cause lost messages.
With RDS the routing rules can be changed dynamically without creating new listener and the already established connections can  use the new rules immediately, so it is a very good feature.

I know the major protocol in envoy is http, but RDS support in other protocols would could come handy sometimes.
For example in our case. Because the lack of it in thrift we had to add hacks on client side, like to poll envoy config dump to check if new rules have applied yet and do reconnect.

On api level the RDS is independent from the listener's protocol. I added the 'message Rds' to the thrift part just because I did not want to cross refer to envoy.extensions.filters.network.http_connection_manager.v3 for the same.
Also the RdsRouteConfigSubscription::onConfigUpdate gets generic Protobuf::Message-s.

Of course I did not want Thrift listener with HTTP routing config. 
The problem with the classes in source/common/router/rds_impl.* is they directly refer to the envoy::config::route::v3::RouteConfiguration and Envoy::Router::Config which are HTTP routing config. So pointless to derive from them.
That is why I just copied them, deleted the unnecessary parts and change the HTTP route config class names to the Thrift counterpart. Even I don't like this solution that much.

What may be better is to turn to templates those classes what I added to the thrift with the protocol's routing config as template parameter and move back to source/common/routing.
These could be used as base class of current HTTP rds classes and with minimal new code possible to introduce RDS support into other protocols as well.

rgs1 commented:
> I'm not sure I fully understand what is proposed. How can Thrift make use of RDS, which is about HTTP routes?
> 
> We definitely should not fork the RDS code, there is a lot of subtlety and complexity in the warming, shared providers and lifecycle (for on-demand). I would recommend making the base code more generic if it doesn't add too much complexity to Envoy core (i.e. it's just about making some resource type opaque).
> 

Yeah, that's what I was referring to -- reusing as much as possible as long as it doesn't become too complex.

rgs1 commented:
> If the static routing rules change in a listener a new listener is created and after a drain period the old one is deleted. This has two negative consequences:
> 
> * During the drain period the clients connected to old listener are using the outdated routing rules.
> * When the drain period is over the connections will be closed by envoy which can cause lost messages.
> 
> With RDS the routing rules can be changed dynamically without creating new listener and the already established connections can use the new rules immediately, so it is a very good feature.
> 
> I know the major protocol in envoy is http, but RDS support in other protocols would could come handy sometimes.
> For example in our case. Because the lack of it in thrift we had to add hacks on client side, like to poll envoy config dump to check if new rules have applied yet and do reconnect.

I agree with all of this, this is a great addition.

> 
> On api level the RDS is independent from the listener's protocol. I added the 'message Rds' to the thrift part just because I did not want to cross refer to envoy.extensions.filters.network.http_connection_manager.v3 for the same.

Makes sense.

> Also the RdsRouteConfigSubscription::onConfigUpdate gets generic Protobuf::Message-s.
> 
> Of course I did not want Thrift listener with HTTP routing config.
> The problem with the classes in source/common/router/rds_impl.* is they directly refer to the envoy::config::route::v3::RouteConfiguration and Envoy::Router::Config which are HTTP routing config. So pointless to derive from them.
> That is why I just copied them, deleted the unnecessary parts and change the HTTP route config class names to the Thrift counterpart. Even I don't like this solution that much.
> 

Yeah that was my concern.

> What may be better is to turn to templates those classes what I added to the thrift with the protocol's routing config as template parameter and move back to source/common/routing.
> These could be used as base class of current HTTP rds classes and with minimal new code possible to introduce RDS support into other protocols as well.

That would be great, shall we go down that path? 

Thanks!

htuch commented:
So, there is probably two things here to consider API and config plane wise:
1. I agree that we probably want the moral equivalent of route providers and RDS for non-HTTP routing. We should be able to reuse almost all of what was created for RDS by making it generic. I would be fairly strongly opposed to copy+paste forking it.
2. We shouldn't use the RDS gRPC endpoint for the same purpose, since it's actually delivering a different xDS resource type. I would suggest when it comes to gRPC endpoint (if not using ADS), that we have a distinct endpoint.

tkovacs-2 commented:
> 
> So, there is probably two things here to consider API and config plane wise:
> 
>     1. I agree that we probably want the moral equivalent of route providers and RDS for non-HTTP routing. We should be able to reuse almost all of what was created for RDS by making it generic. I would be fairly strongly opposed to copy+paste forking it.

Understood.

> 
>     2. We shouldn't use the RDS gRPC endpoint for the same purpose, since it's actually delivering a different xDS resource type. I would suggest when it comes to gRPC endpoint (if not using ADS), that we have a distinct endpoint.

In case of RDS without ADS to introduce new endpoints maybe not necessary. Please see the change in the commit 3235184. We can just map the new type url-s back to envoy.config.route.v3.RouteConfiguration and we can reuse the RouteDiscoveryService endpoint. The DiscoveryRequest/DiscoveryResponse doesn't depend on the protocol also it is backward compatible from xds server point of view.

Alternatively we can say the non-http RDS can be used just with ADS.

tkovacs-2 commented:
> 
> That would be great, shall we go down that path?
> 
> Thanks!

Sure, let

markdroth commented on api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto:
```diff
@@ -92,6 +93,9 @@ message ThriftProxy {
   google.protobuf.UInt32Value max_requests_per_connection = 7;
+
+  // Use xDS to fetch the route configuration. If both route_config and rds defined, rds has priority.
+  Rds rds = 8;
```
Probably want to use a oneof migration annotation here, so that we know to put this field and the `route_config` field in a oneof in the next major version (if that ever happens).

tkovacs-2 replied:
Sorry, it is not clear. What do you mean 'it's not obvious to me that we want to use exactly the same resource type for HTTP RouteConfiguration and Thrift RouteConfiguration'? And 'two resource types happen to be the same right now, who's to say that they won't diverge in the future'?

As XDS resource the HTTP RouteConfiguration and Thrift RouteConfiguration are different for the two different protocols and cannot be exchanged.
But technically the GRPC endpoint RouteDiscoveryService can be used to transfer both. And they are both routing config so we don't try to force something unrelated into it.
I just meant technically it is not needed to introduce new endpoint like let's say RouteDiscoveryServiceThrift for Thrift RouteConfiguration. In type_to_endpoint.cc the endpoint is selected based on the resource type_url with that assumption there is 1:1 mapping between endpoints and resource urls. With this 'alias' we can say: ok use the same endpoint to this resource type like that resource type.  Maybe not the most beautiful but simple and no need to touch this part again if we introduce support for RDS for other protocols.

Yes, in the first version I used 'oneof'. But then that extreme but previously valid config where neither 'route_config' nor 'rds' specified becomes invalid.

markdroth replied:
> What do you mean 'it's not obvious to me that we want to use exactly the same resource type for HTTP RouteConfiguration and Thrift RouteConfiguration'? And 'two resource types happen to be the same right now, who's to say that they won't diverge in the future'?

What I'm saying is that the resource type `envoy.config.route.v3.RouteConfiguration` was designed to contain routing information for HTTP specifically.  Even if thrift happens to need exactly the same format today, it's not clear to me that it necessarily makes sense to use the same resource type directly here, because we don't want to be in a position where in the future we need to either add a bunch of thrift-specific fields to the existing `envoy.config.route.v3.RouteConfiguration` proto or we need to break thrift routing by changing what resource type it uses after everyone is already expecting to use the current resource type.

Alternatively, though, if we can migrate the existing `envoy.config.route.v3.RouteConfiguration` proto to be a more generic form (possibly via the work that @snowp is doing in #17633), then our story can be that that resource type is not actually HTTP-specific, and we have enough extensibility to make it work for any resource type.  I'd like @mattklein123 and @snowp to weigh in on that option.

I think the discussion about the RPC service (use the existing `RouteDiscoveryService` vs. a new one) is actually just a side question; the main issue is what resource type we use here.  The resource type generally maps 1:1 to the RPC service, so if we go with the same resource type, then it makes sense to use the same RPC service, and if we go with a different resource type, then it makes sense to add a new RPC service.

> Yes, in the first version I used 'oneof'. But then that extreme but previously valid config where no 'route_config' specified becomes invalid.

Using a `oneof` does not make the fields in the `oneof` required.  If both `route_config` and `rds` were in the same `oneof`, it would still be legal to specify neither of them.

htuch replied:
@tkovacs-2 the Envoy-side code changes are great, thanks for the refactoring, it's much cleaner.

From what I can see above, we actually have two distinct resources for Thrift and Envoy RDS, we're not reusing `RouteConfiguration` here. I am a bit confused about the "resource alias" annotation though, that seems to imply some connection which Mark points out doesn't have any reasonable association in the API.

markdroth reviewed (commented):
I agree with @htuch that it's not obvious to me that we want to use exactly the same resource type for HTTP RouteConfiguration and Thrift RouteConfiguration.  The resource aliasing thing seems a little ugly.  And even if the two resource types happen to be the same right now, who's to say that they won't diverge in the future?

Also, in terms of the API representation, you may want to use the [new matching API](https://docs.google.com/document/d/1G4g-6q0IArz_ERgqixzZCM0-wbexFYuBUvT9hkV7QRU/edit), which @snowp is working on integrating into RDS in #17633.

tkovacs-2 commented:
@rgs1 @htuch @markdroth 
I have eliminated the copy-paste. Could you please check the latest state?

Also one question: In RouteConfigProviderManagerImpl::createRdsRouteConfigProvider is it neccessary to call init_manager.add() after RdsRouteConfigSubscription creation or it could be done just after RdsRouteConfigProviderImpl creation?

markdroth commented:
This still looks like it's referring to RDS.  I don't think that's right.  If this is not using the resource type `envoy.config.route.v3.RouteConfiguration`, then it should not use RDS; it should have a different RPC service.

There should be no need for the `resource_alias` option here.

tkovacs-2 commented:
@markdroth
I added the Thrift to the method names in trds.proto. As I have seen method of other services named based on similar pattern.
Could you double check and approve the api change again?

markdroth commented:
/lgtm api

htuch commented:
@adisuissa can you do first pass on implementation? Thanks.

tkovacs-2 commented:
/retest

htuch commented:
@tkovacs-2 thanks. I think the next step is as follows:
* @adisuissa will provide a detailed review pass as first-pass reviewer.
* I think we need to break the PR up into probably two PRs. The first should be the RDS and route config provider refactor (i.e. with no Thrift code whatsoever) and the second will be the incremental Thrift usage of RDS. We can refer back to this PR during the review to show how they integrate.
* It's still a little unclear to me how this sits with the efforts in https://github.com/envoyproxy/envoy/pull/18823. I imagine they must have to solve this same problem somehow, @wbpcode @zhaohuabing can you take a look at this PR and comment on how much similarity there is? Have you built an efficient shared route config provider as done in this PR?

zhaohuabing commented:
@htuch We have built an unified RDS solution in Aeraki https://github.com/aeraki-framework/meta-protocol-proxy, proven working with both Thrift and Dubbo. The MetaProtocol PR is intended to push what have done in Aeraki to  the Envoy, including unified RDS, tracing, logging, etc.

tkovacs-2 commented:
Hi @zhaohuabing,
The RDS in Aeraki is a copy-paste of the current RDS implementation in envoy. First I did the same for Thrift RDS in this PR. But it was not accepted by reviewers.
Now the major part of the PR is to make the RDS implementation reusable for other protocols than HTTP. So I suppose this will be useful for you as well to avoid copy-paste.
The Thrift RDS has become a minor part.

One things are not clear for me in the metaprotocol:
What is your plan with current thrift, dubbo, etc. implementations in envoy?
If your aim is to simplify the filter implementations then maybe better approach is to extract the common parts of the current implementations to reusable classes.
Now it seems to me you are building up a parallel universe with the metaprotocol.

zhaohuabing commented:
> Hi @zhaohuabing, The RDS in Aeraki is a copy-paste of the current RDS implementation in envoy. First I did the same for Thrift RDS in this PR. But it was not accepted by reviewers. Now the major part of the PR is to make the RDS implementation reusable for other protocols than HTTP. So I suppose this will be useful for you as well to avoid copy-paste. The Thrift RDS has become a minor part.
> 
The RDS(we call it MetaRDS) in Aeraki is a unified solution for all the applications built on top of the MetaProtocol Proxy, including Thrift, Dubbo, and other proprietary protocols.  We didn't include HTTP/gRPC routing in the MetaRDS because HTTP Routing is already handled very well by the current RDS and HCM, and it's not in the design scope of the MetaProtcol proxy. (Probably we can discuss about that)


> One things are not clear for me in the metaprotocol: What is your plan with current thrift, dubbo, etc. implementations in envoy? If your aim is to simplify the filter implementations then maybe better approach is to extract the common parts of the current implementations to reusable classes. Now it seems to me you are building up a parallel universe with the metaprotocol.

The goal of the MetaProtocol Proxy is to build a unified solution for all request/response, stateless protocols, avoiding duplication and allowing users to create protocol filters more easily. More about the goal and design can be found here: https://docs.google.com/document/d/1y_ALyjGp_H1_TRrrRxZxnqENISqeZI6eLn71FAFsZPk/edit#heading=h.ic0mhtrvlmy9

tkovacs-2 commented:
@htuch
Please see the PR [#18846](https://github.com/envoyproxy/envoy/pull/18846) for the RDS refactor part.

tkovacs-2 commented:
@zhaohuabing 
Yes, I understand the document, but the answer for my question is still not clear.
I suppose you want to merge the thrift (and dubbo) codec together with the metaprotocol filter, at least as example.
What will happen with the current native thrift (and dubbo) filter?
To drop them is a little drastic because the new api is different, also it should provide at least the same features and performance as the native. To keep them is also strange because then these protocols will be duplicated.
The aim was to simplify the filter implementations. So I would not introduce the metaprotocol on api level as a new filter but as a kind of framework for filter implementation. Also turn the existing native filter's code to use the metaprotocol framework with their current api.
To make a new filter api is not a big thing, so I don't feel the unified api as big gain, but rather limit on long term when the needs of different protocols start to diverge. 
But this is just my personal opinion.

zhaohuabing commented:
> @zhaohuabing 
> Yes, I understand the document, but the answer for my question is still not clear.
> I suppose you want to merge the thrift (and dubbo) codec together with the metaprotocol filter, at least as example.
> What will happen with the current native thrift (and dubbo) filter?
> To drop them is a little drastic because the new api is different, also it should provide at least the same features and performance as the native. To keep them is also strange because then these protocols will be duplicated.

The framework code will be pushed to the MetaProtocol proxy filter. Although Dubbo and Thrift have been proven working on top of the Meta Protocol proxy, however, we need community consense if we'd like to port dubbo and thrift to metaprotocol. The core maintainers and owners of Dubbo and Thrift should agree on that before the shift. I personally think it's not the key issue here, Thrift filter could exist as it is now if the community think it should be, and the MetaProtocol proxy can be used to facilitate protocols that are not currently supported in the Envoy.

> The aim was to simplify the filter implementations. So I would not introduce the metaprotocol on api level as a new filter but as a kind of framework for filter implementation. Also turn the existing native filter's code to use the metaprotocol framework with their current api.
> To make a new filter api is not a big thing, so I don't feel the unified api as big gain, but rather limit on long term when the needs of different protocols start to diverge. 
> But this is just my personal opinion.


The implementation will be pushed in another PR after we agreed on the design and the API.

wbpcode commented:
@tkovacs-2 We hope that the meta protocol proxy will make it easier to implement new protocols in Envoy, including various private protocols. So we will provide a full set of proto APIs and a new implementation.
**Replacing the dubbo proxy and thrift proxy is not our goal.** Unless the meta protocol proxy proves to work very well in Envoy in the future, and the community agrees, we may replace their implementations while keeping the existing dubbo/thrift APIs the same.
**Since this is a lot of work. So we split the PR into multiple, proto APIs, core frameworks, features, etc.**

htuch commented:
I think there are a few things going on here that would be good to clarify:
1. @zhaohuabing @wbpcode the point that @tkovacs-2 makes around the MetaRDS implementation has me a little concerned. Can you point me to the MetaRDS implementation? We've been working with @tkovacs-2 so that we don't end up doing a copy+paste of the common route config provider code from the core in this PR, and I expect we might do the same for MetaRDS. If so, by first landing https://github.com/envoyproxy/envoy/pull/18846, we can maybe be in a better position to integrate MetaRDS without as much copy+paste from how RDS works today (if that is how it is done). Do you think https://github.com/envoyproxy/envoy/pull/18846 would help with MetaRDS?
2. @zuercher @rgs1 can you weight in on the positioning of the existing Thrift filter with the MetaProtocol Thrift? I agree it's confusing to have two independent implementations and we'll want some clarity on which is the production ready solution.

wbpcode commented:
Hi, @htuch We will rewrite the whole meta protocol proxy, so it doesn't matter what the old implementation is before. After merging the core framework of new meta protocol proxy, we will have another special PR and design to do this work. I think at that time, #18846 will be very helpful. 

And, replacing the dubbo proxy and thrift proxy is not our goal. Although we have implemented thrift codec/dubbo codec in the old meta protocol proxy, but it would not be the part of new meta protocol proxy. So there is no need to worry about the co-existence of thrift proxy and meta protocol thrift.

tkovacs-2 commented:
@wbpcode 
Ok, clear.

@htuch 
>     1. #18846 would help with MetaRDS?
Yes, the current MetaRDS implementation I think is [here](https://github.com/aeraki-framework/meta-protocol-proxy/tree/master/src/meta_protocol_proxy/route) (rds.h rds_impl.* route_config_*)
Once the #18846 is merged, basically it can be the same like the Thrift RDS part in this PR.

htuch commented:
OK, thanks for the clarity; it makes sense to hold back on replacing extant filters until we have stability here and there is enough commonality that we can ahead with the refactoring PR. I wonder if we want both TRDS and MetaRDS though; could these be the same t hing once we merge #18846?

tkovacs-2 commented:
I think they can't be exactly the same, since the type of route config (`ConfigImpl` and the related proto) is different for them.
And somebody need to create these objects of the proper types. That means filter specific implementations of `RouteConfigProviderManagerImpl` and `ConfigTraitsImp` will always be needed.
However in case of TRDS, MetaRDS and other simple cases (when the `ConfigImpl` depends only on the proto) the difference is only the types names and the code is just copy-paste.

I have planned to introduce a template default implementation for these under `Envoy::Rds`.
In this way the code what is currently in `thrift_proxy/router/rds.h`, `rds_impl.h`, `rds_impl.cc` and `rds_impl_test.cc` just need once. Otherwise same would be needed for every filter.

zuercher reviewed (commented):
Thanks! I think this looks good generally, but I was only able to give it cursory look today. I'm also out tomorrow, so I can't get back to this until Friday.

tkovacs-2 commented:
@zuercher 
Thanks for your comments. For easier review meanwhile I split the generic part into a separate PR #18846. I will add the fixes for it there.

@htuch Do you have some suggestion how to handle thrift part?
I mean that depends on the generic part. If I merge the branch of #18846 into the branch of this PR, can github handle that? 
Or just collect the comments here and fix here after #18846 is merged?

tkovacs-2 commented on source/common/rds/rds_route_config_subscription.cc:
```diff
@@ -0,0 +1,126 @@
+    Server::Configuration::ServerFactoryContext& factory_context, const std::string& stat_prefix,
+    RouteConfigProviderManager& route_config_provider_manager)
+    : route_config_name_(route_config_name),
+      scope_(factory_context.scope().createScope(stat_prefix + "rds." + route_config_name_ + ".")),
```
> I think it's probably ok for trds stats to have rds in their name, the stat prefix should make clear which is which
Yes, exactly.
I wanted to use the same logic for naming on api. The routing is never used in itself, it is always related to a filter. So some package name, namespace, prefix, etc. is there to make clear what kind of rds we are talking about.
Originally I wanted to name `envoy.service.thrift.v3.ThriftRouteDiscoveryService` and `envoy.extensions.filters.network.thrift_proxy.v3.Trds` without `Trift` and 'T`. But I was alone with that view.
For me rds is generic but seems for others rds == http. 

For logging 'Logger::Id::router' is used now but that is obviously wrong. I introduce a new id like 'Logger::Id::rds'. I'm afraid  not possible to further specify without bring back templates at the generic classes:
`class RdsRouteConfigSubscription : Envoy::Config::SubscriptionCallbacks, protected Logger::Loggable<Logger::Id::router>`

For init manager do you mean these names should contain some filter specific?
`parent_init_target_(fmt::format("RdsRouteConfigSubscription init {}", route_config_name_) ...
local_init_watcher_(fmt::format("RDS local-init-watcher {}", route_config_name_) ...
local_init_target_(fmt::format("RdsRouteConfigSubscription local-init-target {}", route_config_name_) ...
local_init_manager_(fmt::format("RDS local-init-manager {}", route_config_name_))`

There is just one exception from in the generic part:
`throw EnvoyException(fmt::format("Unexpected RDS configuration ...`

I can reuse the stat_prefix in these strings to make difference.
Or I can introduce a new parameter for `RdsRouteConfigSubscription` to use instead the current hardcoded "RDS" or "rds" literals. Which one do you prefer?

zuercher replied:
> For logging 'Logger::Id::router' is used now but that is obviously wrong. I introduce a new id like 'Logger::Id::rds'.

I think that's fine.

> For init manager do you mean these names should contain some filter specific?

I think these should report TRDS vs. RDS. I expect someone will eventually use the same route config and maybe even stat prefix name for both http and thrift rds and these debugging statements won't help in that case. I think it makes sense to parameterize these. Perhaps just a parameter for "RDS" vs "TRDS" and add that string somewhere in the messages?

tkovacs-2 replied:
Fixed in #18846

htuch commented:
@tkovacs-2 "Or just collect the comments here and fix here after #18846 is merged?" <-- yep, GitHub doesn't have very good support for a PR pipeline that I know of.

rojkov commented:
Wait for #18846 to be merged.
/wait

tkovacs-2 commented on envoy/rds/route_config_provider.h:
```diff
@@ -0,0 +1,63 @@
+
+    // The discovery version that supplied this route. This will be set to "" in the case of
+    // static clusters.
+    std::string version_;
```
Fixed in #18846

tkovacs-2 commented:
/reopen

tkovacs-2 commented on source/extensions/filters/network/thrift_proxy/router/rds.h:
```diff
@@ -0,0 +1,66 @@
+
+/**
+ * The RouteConfigProviderManager exposes the ability to get a RouteConfigProvider. This interface
+ * is exposed to the Server's FactoryContext in order to allow HttpConnectionManagers to get
```
Done

tkovacs-2 commented on api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto:
```diff
@@ -92,6 +105,9 @@ message ThriftProxy {
   // Optional maximum requests for a single downstream connection. If not specified, there is no limit.
   google.protobuf.UInt32Value max_requests_per_connection = 7;
+
+  // Use xDS to fetch the route configuration. If both route_config and trds defined, trds has priority.
```
Done
